mmengine/runner/deepspeed_runner.py

```diff
@@ -0,0 +1,771 @@
+
+        # Collect and log environment information.
+        self._log_env(env_cfg)
+
+        # collect information of all modules registered in the registries
+        registries_info = count_registered_modules(
+            self.work_dir if self.rank == 0 else None, verbose=False)
+        self.logger.debug(registries_info)
+
+        # Build `message_hub` for communication among components.
+        # `message_hub` can store log scalars (loss, learning rate) and
+        # runtime information (iter and epoch). Those components that do not
+        # have access to the runner can get iteration or epoch information
+        # from `message_hub`. For example, models can get the latest created
+        # `message_hub` by
+        # `self.message_hub=MessageHub.get_current_instance()` and then get
+        # current epoch by `cur_epoch = self.message_hub.get_info('epoch')`.
+        # See `MessageHub` and `ManagerMixin` for more details.
+        self.message_hub = self.build_message_hub()
+        # visualizer used for writing log or visualizing all kinds of data
+        self.visualizer = self.build_visualizer(visualizer)
+
+        self._load_from = load_from
+        self._resume = resume
+        # flag to mark whether checkpoint has been loaded or resumed
+        self._has_loaded = False
+
+        # load DeepSpeed configuration file
+        self.ds_config = self.cfg.get('ds_config', None)
+        assert self.ds_config is not None, 'ds_config should be specified.'
+
+        self.check_ds_config(self.ds_config)
+
+        # build a model
+        if isinstance(model, dict) and data_preprocessor is not None:
+            # Merge the data_preprocessor to model config.
+            model.setdefault('data_preprocessor', data_preprocessor)
+        self.model = self.build_model(model)
+
+        # initialize the model weights before wrapping it with deepspeed
+        self._weights_initialized = False
+        self._init_model_weights()
+
+        # wrap model
+        self.model = self.wrap_model(
+            self.cfg.get('model_wrapper_cfg'), self.model)
+
+        # build optim wrapper
+        self.optim_wrapper = self.build_optim_wrapper(self.optim_wrapper)
+        assert isinstance(self.optim_wrapper, DeepSpeedOptimWrapper), (
+            'OptimWrapper type should be \'DeepSpeedOptimWrapper\' when using '
+            'DeepSpeedRunner.')
+
+        # initialize DeepSpeed Engine
+        self.model, optimizer, _, _ = deepspeed.initialize(
+            model=self.model,
+            optimizer=self.optim_wrapper.optimizer,
+            model_parameters=self.model.parameters(),
+            config=self.ds_config)
+        self.optim_wrapper.optimizer = optimizer
+
+        # Set logging level to remove duplicate training log from DeepSpeed
+        deepspeed_logger = logging.getLogger('DeepSpeed')
+        deepspeed_logger.setLevel(logging.WARNING)
+
+        self.inject_basemodel_methods()
```

Comment:
Can I ask you to provide some links to those examples?
One of the difficulties to handle this is that `DeepSpeedEngine` fails to get some attributes from `self.module`. https://github.com/microsoft/DeepSpeed/blob/90ae6884424232870154b49967c3e61f0db550d6/deepspeed/runtime/engine.py#L461

I think it will be difficult to support them in the current implementation. Maybe we have find the other way to handle this.

Reply:
Some low-level tasks need to rewrite `train_step`, such as [GAN in mmediting](https://github.com/open-mmlab/mmediting/blob/1.x/mmedit/models/base_models/base_gan.py#L420). It is indeed very difficult to support them, so I think current implementation is acceptable.

Reply:
@nijkah 

Maybe this one?
https://github.com/open-mmlab/mmediting/blob/dev-1.x/docs/en/advanced_guides/1_models.md